subsys/sensing/sensor_mgmt.c

```diff
@@ -0,0 +1,383 @@
+		__ASSERT(reporter, "sensor's reporter should not be NULL");
+
+		init_connection(conn, reporter, sensor);
+
+		LOG_DBG("init sensor, reporter:%s, client:%s, connection:%d",
+			reporter->dt->dev->name, sensor->dt->dev->name, i);
+
+		/* link connection to its repoter's client_list */
+		sys_slist_append(&reporter->client_list, &conn->snode);
+
+		conns[i] = conn;
+	}
+
+	/* physical sensor is working at polling mode by default,
+	 * virtual sensor working mode is inherited from its reporter
+	 */
+	if (is_phy_sensor(sensor)) {
+		sensor->mode = SENSOR_TRIGGER_MODE_POLLING;
+	}
+
+	return sensor_api->init(sensor->dt->dev, &sensor->dt->info, conns, conns_num);
+}
+
+/* create struct sensing_sensor *sensor according to sensor device tree */
+static struct sensing_sensor *create_sensor(struct sensing_dt_info *dt)
+{
+	struct sensing_sensor_ctx *sensor_ctx;
+	struct sensing_sensor *sensor;
+	uint16_t sample_size, total_size;
+	uint16_t conn_sample_size = 0;
+	uint16_t i = 0;
+	void *tmp_data;
+
+	__ASSERT(dt && dt->dev, "device tree info or sensing dev is invalid");
+	sensor_ctx = dt->dev->data;
+	__ASSERT(sensor_ctx, "sensing sensor context is invalid");
+
+	sample_size = sensor_ctx->register_info->sample_size;
+	for (i = 0; i < dt->reporter_num; i++) {
+		conn_sample_size += get_sensor_sample_size_from_dt(dt, i);
+	}
+
+	/* total memory to be allocated for a sensor according to sensor device tree:
+	 * 1) size of struct sensing_sensor *
+	 * 2) sample data point to struct sensing_sensor->data_buf
+	 * 3) size of struct sensing_connection* for sensor connection to its reporter
+	 * 4) reporter sample size to be stored in connection data
+	 */
+	total_size = sizeof(*sensor) + sample_size +
+		     dt->reporter_num * sizeof(*sensor->conns) + conn_sample_size;
+
+	/* total size for different sensor maybe different, for example:
+	 * there's no reporter for physical sensor, so no connection memory is needed
+	 * reporter num of each virtual sensor may also different, so connection memory is also
+	 * varied, so here malloc is a must for different sensor.
+	 */
+	sensor = malloc(total_size);
+	if (!sensor) {
+		LOG_ERR("malloc memory for struct sensing_sensor error");
+		return NULL;
+	}
+	sensor->sample_size = sample_size;
+	sensor->data_buf = (uint8_t *)sensor + sizeof(*sensor);
+	sensor->conns = (struct sensing_connection *)((uint8_t *)sensor->data_buf + sample_size);
+
+	tmp_data = sensor->conns + dt->reporter_num;
+	for (i = 0; i < dt->reporter_num; i++) {
+		sensor->conns[i].data = tmp_data;
+		tmp_data = (uint8_t *)tmp_data + get_sensor_sample_size_from_dt(dt, i);
+	}
+
+	__ASSERT(tmp_data == ((uint8_t *)sensor + total_size), "sensor memory assign error");
+
+	sensor->dt = dt;
+	sensor->dt->info.flags = sensor_ctx->register_info->flags;
+	sensor->dt->info.version = sensor_ctx->register_info->version;
+
+#ifdef CONFIG_SENSOR_INFO
+	STRUCT_SECTION_FOREACH(sensor_info, tmp) {
+		if (!strcmp(sensor->dt->dev->name, tmp->dev->name)) {
```

Comment:
This should compare the `dev` pointer directly instead of doing a string compare of the name.

Reply:
modified to dev compare directly, thanks.